src/main/java/com/mparticle/kits/PilgrimSdkKit.java

```diff
@@ -0,0 +1,199 @@
+
+import com.foursquare.pilgrim.LogLevel;
+import com.foursquare.pilgrim.PilgrimSdk;
+import com.foursquare.pilgrim.PilgrimUserInfo;
+import com.mparticle.MParticle;
+import com.mparticle.consent.ConsentState;
+import com.mparticle.identity.MParticleUser;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This is the PilgrimSdk mParticle kit , used to extend the functionality of mParticle SDK and allow control of
+ * the PilgimSdk; Mapping analogous public mParticle APIs into PilgrimSdk's API.
+ * <p>
+ * <p>
+ * In addition to this file, you also will need to edit:
+ * - ./build.gradle (as explained above)
+ * - ./README.md
+ */
+final public class PilgrimSdkKit extends KitIntegration implements KitIntegration.UserAttributeListener, KitIntegration.IdentityListener {
+
+    /**
+     * Name of the kit
+     */
+    private static final String KIT_NAME = "PilgrimSdkKit";
+
+    /**
+     * Key used to get the Pilgrim Sdk api key from mParticle's settings response
+     */
+    static final String SDK_KEY = "pilgrim_sdk_key";
+
+    /**
+     * Key used to get the Pilgrim Sdk secret from mParticle's settings response
+     */
+    static final String SDK_SECRET = "pilgrim_sdk_secret";
+
+    /**
+     * Key used to get the Pilgrim Sdk enable  configuration enabling flag from mParticle's settings response
+     */
+    private static final String SDK_ENABLE_PERSISTENT_LOGS = "pilgrim_persistent_logs";
+
+    private static final String MPARTILE_USER_ID = "mParticleUserId";
+
+
+    @Override
+    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
+        String key = settings.get(SDK_KEY);
+        if (KitUtils.isEmpty(key)) {
+            throw new IllegalArgumentException("PilgrimSdk key is empty.");
+        }
+
+        String secret = settings.get(SDK_SECRET);
+        if (KitUtils.isEmpty(secret)) {
+            throw new IllegalArgumentException("PilgrimSdk secret is empty.");
+        }
+
+        PilgrimSdk.Builder builder = new PilgrimSdk.Builder(context)
+                .consumer(key, secret)
+                .logLevel(LogLevel.DEBUG);
+
+        if (KitUtils.parseBooleanSetting(settings, SDK_ENABLE_PERSISTENT_LOGS, false)) {
+            builder.enableDebugLogs();
+        }
+
+        // Configure with our starter
+        PilgrimSdk.with(builder);
+
+        List<ReportingMessage> messageList = new LinkedList<>();
+        // Can we add messages to track if initialized/started successfully in attributes
+        messageList.add(new ReportingMessage(this, ReportingMessage.MessageType.APP_STATE_TRANSITION, System.currentTimeMillis(), null));
+        return messageList;
+    }
+
+    @Override
+    public String getName() {
+        return KIT_NAME;
+    }
+
+    @Override
+    public List<ReportingMessage> setOptOut(boolean optedOut) {
+        List<ReportingMessage> messageList = new LinkedList<>();
+        PilgrimSdk.stop(getContext());
```

Comment:
Here you are likely going to want to check if `optedOut ==true` before stopping the Pilgrim SDK. This method may be invoked even if the value isn't a change, so you could be getting `optedOut == false` even if the current optOut state already false, in which case, you would not want the SDK to stop

Reply:
Sounds good, That wasn't specified in the documentation I found so I figured it was just called when opting out once.